crypto-ffi/src/wasm/mod.rs

```diff
@@ -79,96 +79,201 @@ pub extern "C" fn __stack_chk_fail() {
+        match value {
+            core_crypto::ProteusError::ProteusSessionError(SessionError::InternalError(
+                proteus_wasm::internal::types::InternalError::NoSessionForTag,
+            )) => Self::SessionNotFound,
+            core_crypto::ProteusError::ProteusSessionError(SessionError::DuplicateMessage) => Self::DuplicateMessage,
+            core_crypto::ProteusError::ProteusSessionError(SessionError::RemoteIdentityChanged) => {
+                Self::RemoteIdentityChanged
+            }
+            _ => Self::Other(value.error_code()),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error, strum::AsRefStr)]
+pub(crate) enum InternalError {
     #[error(transparent)]
-    CryptoError(#[from] CryptoError),
+    MlsError(#[from] MlsError),
+    #[cfg(feature = "proteus")]
     #[error(transparent)]
-    E2eError(#[from] E2eIdentityError),
+    ProteusError(#[from] ProteusError),
+    #[error("End to end identity error")]
+    E2eiError,
     #[error(transparent)]
     SerializationError(#[from] serde_wasm_bindgen::Error),
-    #[error("Failed lifting an enum")]
-    EnumError,
+    #[error("Unknown cyphersuite identifier")]
+    UnknownCyphersuite,
     #[error("Transaction rolled back. Last proteus error code: {proteus_error_code:?}. Uncaught JsError: {uncaught_error:?}")]
     TransactionFailed {
         uncaught_error: JsValue,
-        proteus_error_code: u32,
+        proteus_error_code: Option<u16>,
     },
 }
 
-#[derive(Debug, Clone, serde::Serialize)]
-#[serde(rename_all = "camelCase")]
-struct CoreCryptoJsRichError {
-    error_name: String,
-    message: String,
-    rust_stack_trace: String,
-    proteus_error_code: u32,
-}
-
-impl<'a> From<&'a CoreCryptoError> for CoreCryptoJsRichError {
-    fn from(e: &'a CoreCryptoError) -> Self {
-        Self {
-            error_name: match e.0 {
-                WasmError::CryptoError(_) => "CryptoError",
-                WasmError::E2eError(_) => "E2eError",
-                WasmError::SerializationError(_) => "SerializationError",
-                WasmError::EnumError => "EnumError",
-                WasmError::TransactionFailed { .. } => "TransactionFailed",
+// This implementation is intended to be temporary; we're going to be completely restructuring the way we handle
+// errors in `core-crypto` soon. We can replace this with better error patterns when we do.
+//
+// Certain error mappings could apply to both MLS and Proteus. In all such cases, we map them to the MLS variant.
+// When we redesign the errors in `core-crypto`, these ambiguities should disappear anyway.
+impl From<CryptoError> for InternalError {
+    fn from(value: CryptoError) -> Self {
+        #[cfg(feature = "proteus")]
+        if let Some(error_code) = value.proteus_error_code() {
+            return match error_code {
+                102 => ProteusError::SessionNotFound, // reviewers take note: is this correct?
```

Comment:
Yes, I would say this corresponds to what's on
https://wearezeta.atlassian.net/wiki/spaces/ENGINEERIN/pages/689045561/Proteus+decryption+error+code+explanation

and the code in proteus repo, so feel free to remove the `reviewers ...` comment.